ships/0005-local-source-upload.md

```diff
@@ -0,0 +1,262 @@
+the actual build process, while we already have plans to support data upload towards a
+[container-registry][LocalSourceCodeEP] alternatively.
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][KubectlCP]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][KubectlExec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the /workspace/source mount point for all containers in the build-pod. Tekton also
+adds it's own helper to ["git clone" the repository][ControllerGitClone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][ControllerWaiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][ControllerWaiterUsage].
+
+##### BuildRun Changes
+
+The proposed API changes are the following:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata: {}
+spec:
+  sources:
+    - name: src
+      type: LocalCopy
+```
+
+The Build Controller will inspect for a type `LocalCopy` on the "`sources"` slice, and therefore
+instead of adding the git clone step, it will instead wait for the user to upload.
+
+The `sources` slice is part of [Remote Artifacts (EP)][RemoteArtifactsEP] and has been worked on, so
+more effort may be needed to create a good API fit.
+
+#### Waiting Mechanism
+
+The container-image built by Shipwright is created by a series of steps running sequentially, those
+steps are executed in the build-pod, running each container and inspecting for return-code and
+errors.
+
+The task at hand, waiting for the users' data transfer, will follow the same basic structure, which
+means a simple application meant to wait for a specific condition, and exit gracefully afterwards.
+Here we may keep checking if a given "lock" file exists, when removed the application exists. It also
+must exit with an error on timeout.
+
+Additionally, we can employ this application to make verifications on the streamed data, like for
+instance checksum, ownership (UID/GID) and more. Another important duty is to have `tar` installed
+and compatible with the activities performed by the CLI.
+
+The implementation written for research and development can be found at [otaviof/waiter][Waiter]
```

Comment:
My understanding is that for waiting we will just prepend a new `step`, that will replace the `git` one when this feature is desired via CLI.

For the `waiter` image, we will need to place the code under https://github.com/shipwright-io/build/tree/main/cmd, similar to the `git` dir there, so that we can build an image from it, and define test cases under it´s package. This image should be compliant with how we are building images today ( _ubi based_ ), so we need to be able to define this in our [controller YAML](https://github.com/shipwright-io/build/blob/main/deploy/500-controller.yaml#L35-L36) with the `ko` nomenclature.

The `waiter` image should also support the definition of environment variables to modify the container spec where it is used, see our docs on [configuration](https://github.com/shipwright-io/build/blob/main/docs/configuration.md) , in specific the `GIT_CONTAINER_TEMPLATE`.

All of the above needs to be mentioned here.

Reply:
Thanks for the heads up. I've just amended the text to reflect those points, please consider.